andela/ah-the-jedi-backend
# #165305279 user can bookmark or unbookmark an article.

#### What does this PR do?
Enable a user to bookmark or unbookmark an article.
#### Description of Task to be completed?
- Implement `POST`, `GET` and `DELETE` endpoints for bookmarking, getting bookmarked articles and unbookmarking articles respectively.
- Write tests for the implementation.
#### How should this be manually tested?
1. Clone this repo.
   `$ git clone https://github.com/andela/ah-the-jedi-backend.git`
2. Fetch all origin to get all branches
   `$ git fetch origin`
3. Checkout to `ft-user-bookmark-article-165305279` branch.
   `$ git checkout ft-user-bookmark-article-165305279`
4. Activate `virtualenv`
    `$ virtualenv -p python .venv`
    `$ source .venv/bin/activate`
5. Install requirements
   `$ pip install -r requirements.txt`
6. Start server
   `$ python manage.py runserver`
7. Test the endpoint on postman in the following steps:
- Ensure the user is authenticated.
- To bookmark an article - `POST /api/articles/<slug>/bookmark/`
- To unbookmark an article - `DEL /api/articles/<slug>/bookmark/`
- Get the articles you bookmarked - `GET /api/articles/<slug>/bookmark/`
#### What are the relevant pivotal tracker stories?
[#165305279](https://www.pivotaltracker.com/story/show/#165305279)
#### Screenshot
Successful bookmark article
<img width="1097" alt="Screenshot 2019-05-08 at 16 51 42" src="https://user-images.githubusercontent.com/17095461/57380441-abfac800-71b1-11e9-9abc-1d68f8e29388.png">

Successful unbookmark article
<img width="1091" alt="Screenshot 2019-05-08 at 16 52 45" src="https://user-images.githubusercontent.com/17095461/57380496-c0d75b80-71b1-11e9-9ab4-98d64d0f4a08.png">

Get all bookmarked articles
<img width="1097" alt="Screenshot 2019-05-08 at 16 56 52" src="https://user-images.githubusercontent.com/17095461/57380809-5b379f00-71b2-11e9-81e1-3b4dd5b55fd0.png">

A user cannot bookmark an article twice.
<img width="1107" alt="Screenshot 2019-05-08 at 17 58 13" src="https://user-images.githubusercontent.com/17095461/57385321-ecab0f00-71ba-11e9-9582-9f0ffe66ad42.png">


A user cannot bookmark his/her own article.
<img width="1085" alt="Screenshot 2019-05-08 at 16 57 46" src="https://user-images.githubusercontent.com/17095461/57380856-74405000-71b2-11e9-85da-cc50255b7b31.png">
